### ISSUE TITLE
Mocking the custom HOC component during testing

### ISSUE THREAD

**Issue by mussaiin-spotnana** [callstack/react-native-testing-library#794]
## Ask your Question

Hello everybody!
I am new to this testing library so I have a question related to mock implementation of custom HOC components that wrap React-Native components.
 
For example: I created `Typography` component that is making some changes above the `<Text>` core component of react-native. I am using it in `MainScreen.tsx` component and want to test that screen.
And when it comes to the testing, the render function of react-native-testing-library cannot access the texts on the 
`MainScreen` view until I mock the implementation of `Typography` component.

My question is: 
**_The `Typography` component will be used in all cases when we have to render the text on the screen. I think it is not the good way of testing if we have to mock it for every test when we want to access some text. Is there any other solution to make this component accessible for any render of the component during the testing or maybe reusable part of code to mock this component?_**

**Comment by thymikee** [comment]
Can you post a sample implementation of Typography?

**Comment by mussaiin-spotnana** [comment]
@thymikee 
Implementation of Typography:
```
const Typography = ({ style, children, ...props }: TypograpyProps): JSX.Element => {

  // here is the code for changes to the styles of the text, applying fonts and etc. 

  return (
    <Text style={[style, fontWeightStyle]} {...props}>
      {children}
    </Text>
  );
};
```

How I am implementing the mock to access the texts:
```
jest.mock('../../../../shared/Typography', () => jest.fn(() => <></>));
const TypographyMock = Typography as jest.Mock;

describe('...', () => {
  beforeEach(() => {
    ...
    TypographyMock.mockImplementation(({ children, style, testID }) => (
      <Text style={style} testID={testID}>
        {children}
      </Text>
    ));
    ...
  });
  
  test('should allow user to change selected option', () => {
      render(<MainScreen filterValue={filterValue} updateFilter={updateFilter} />);

      const checkboxListProps = CheckboxMock.mock.calls.map(call => call[0]);
      dataArray.forEach(([code, label], index) => {
        expect(checkboxListProps[index]).toMatchObject({
          title: { props: { children: label } },
          checked: filterValue.includes(code),
        });
      });
  });
```
Without this way of mock implementation of Typography render seems to be empty and doesn't contain any texts until I implement the mock of this component.

**Comment by thymikee** [comment]
Why do you want to mock Typography in the first place? In testing we want to avoid mocking our own code because it's omitted from the test coverage. Only mock what's actually necessary like network requests or native modules. If you'd like to simplify Typography for any reason, you'll need to make it use `Text` from `react-native` package, like this (although I don't recommend that as just noted):
```js
jest.mock('../../../../shared/Typography', () => jest.fn(({children}) => <Text>{children}</Text>))
```

**Comment by mussaiin-spotnana** [comment]
Sorry, I think you missed the question. @thymikee 
If we don’t mock the Typography component, the text content is not accessible. 
Testing code looks like this:
```
render(<MainScreen filterValue={filterValue} updateFilter={updateFilter} />);
expect(screen.getByText('Hello world')).toBeInDocument(); // This statement fails
```

Main Screen render code is simply just like this:
```
return (
 <View style={styles.containerStyle}>
    <Typography style={styles.textStyle}>
        Hello world
    </Typography>          
</View>
);
```

**Comment by thymikee** [comment]
If the Text inside Typography is imported from react-native, then it should be visible in the component tree. Please send a minimal repro and an output from the `debug()` call

**Comment by thymikee** [comment]
Also, `.toBeInDocument()` matcher is not available in jest-native: https://github.com/testing-library/jest-native/blob/main/src/index.js nor is it a default Jest matcher

**Comment by mussaiin-spotnana** [comment]
Hi @thymikee!
I've gone through the all my unit tests using texts and some other npm libraries and was wrong about that issue. Sorry for taking your time and thank you a lot!
The tests was failing due to the another issue (structure of the passing data was changed).
I fixed that and then thought that it is related to mocking of custom components, applied the mocks just to be sure and didn't test the first solution separately, but all together. Tests passed successfully and it seemed to me that the issue was strictly in the mock implementation of our custom components.

I thought about that because of the issue that appeared before while I was unit testing usages of my Custom Bottom Drawer component and forgot that for the drawer we were using not core component but separate library @ react-native-modal.

So briefly:
As you answered before, all the core components of react-native are accessible in unit testing, if it is not a separate npm library.
I removed that mock implementations for Typography and tests work fine.
The question is totally closed.
